src/librbd/cache/ParentCacheObjectDispatch.cc

```diff
@@ -0,0 +1,228 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/WorkQueue.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/Journal.h"
+#include "librbd/Utils.h"
+#include "librbd/io/ObjectDispatchSpec.h"
+#include "librbd/io/ObjectDispatcher.h"
+#include "librbd/io/Utils.h"
+#include "librbd/cache/ParentCacheObjectDispatch.h"
+#include "osd/osd_types.h"
+#include "osdc/WritebackHandler.h"
+
+#include <vector>
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::cache::ParentCacheObjectDispatch: " \
+                           << this << " " << __func__ << ": "
+
+using namespace ceph::immutable_obj_cache;
+using librbd::util::data_object_name;
+
+namespace librbd {
+namespace cache {
+
+template <typename I>
+ParentCacheObjectDispatch<I>::ParentCacheObjectDispatch(
+    I* image_ctx) : m_image_ctx(image_ctx), m_cache_client(nullptr),
+    m_initialized(false), m_connecting(false),
+    m_lock("librbd::cache::ParentCacheObjectDispatch::m_lock") {
+  std::string controller_path =
+    ((CephContext*)(m_image_ctx->cct))->_conf.get_val<std::string>("immutable_object_cache_sock");
+  m_cache_client = new CacheClient(controller_path.c_str(), m_image_ctx->cct);
+}
+
+template <typename I>
+ParentCacheObjectDispatch<I>::~ParentCacheObjectDispatch() {
+    delete m_cache_client;
+}
+
+template <typename I>
+void ParentCacheObjectDispatch<I>::init() {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 5) << dendl;
+
+  if (m_image_ctx->parent != nullptr) {
+    ldout(cct, 5) << "child image: skipping" << dendl;
+    return;
+  }
+
+  ceph_assert(m_connecting.load() == false);
+  m_connecting.store(true);
+  Context* create_session_ctx = new FunctionContext([this](int ret) {
+    Mutex::Locker locker(m_lock);
+    m_connecting.store(false);
+  });
+
+  create_cache_session(create_session_ctx, false);
+
+  m_image_ctx->io_object_dispatcher->register_object_dispatch(this);
+  m_initialized = true;
+}
+
+template <typename I>
+bool ParentCacheObjectDispatch<I>::read(
+    uint64_t object_no, uint64_t object_off,
+    uint64_t object_len, librados::snap_t snap_id, int op_flags,
+    const ZTracer::Trace &parent_trace, ceph::bufferlist* read_data,
+    io::ExtentMap* extent_map, int* object_dispatch_flags,
+    io::DispatchResult* dispatch_result, Context** on_finish,
+    Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << "object_no=" << object_no << " " << object_off << "~"
+                 << object_len << dendl;
+  ceph_assert(m_initialized);
+  string oid = data_object_name(m_image_ctx, object_no);
+
+  /* if RO daemon still don't startup, or RO daemon crash,
+   * or session occur any error, try to re-connect daemon.*/
+  if (!m_cache_client->is_session_work()) {
+    {
+      Mutex::Locker locker(m_lock);
```

Comment:
Nit: don't need the mutex if you just use ```if (m_connecting.exchange(true)) {``` since if it returns true it means that it's already re-connecting. Regardless, you would still need to check ```is_session_work(ing)``` before you continue since otherwise you might have raced w/ a non-working connection (line 82) that was restored by another thread by the time you reached here.

Reply:
hi, jason. many thanks for your suggestions. i have refined this code snippet, but i am not sure whether i can totally got your points.

Reply:
Since you don't have a giant lock protecting the full re-connect process, if thread A first passes the ```if (!m_cache_client->is_session_work())```, thread B could have also passed it and reconnected before thread A resumes and executes ```if (!m_connecting.exchange(true))```. This will result in thread A re-connecting a working session. Therefore, you can just add another test after the exchange to see if the session is still broken and needs to be reconnected,

Reply:
done. thanks jason